pkg/platform/aws/aws.go

```diff
@@ -146,6 +147,24 @@ func (c *config) Initialize(ex *terraform.Executor) error {
 		return err
 	}
 
+	// Extract control plane chart files to cluster assets directory.
```

Comment:
I am still against extracting the charts this way. Platform code does not need to know **how** the charts are extracted. It only needs to define **what** charts needs to be defined. I propose following patch to simplify it:
```diff
From 5e9da908110f4d139dadd622764d67ff42ccf3a1 Mon Sep 17 00:00:00 2001
From: Mateusz Gozdek <mateusz@kinvolk.io>
Date: Thu, 2 Jul 2020 13:31:10 +0200
Subject: [PATCH] Simplify extracting controlplane charts

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>
---
 cli/cmd/cluster.go                  |  9 +++++++
 pkg/platform/aws/aws.go             | 30 +++++++----------------
 pkg/platform/baremetal/baremetal.go | 30 +++++++----------------
 pkg/platform/packet/packet.go       | 38 +++++++----------------------
 pkg/platform/platform.go            | 27 ++++++++++++++------
 5 files changed, 55 insertions(+), 79 deletions(-)

diff --git a/cli/cmd/cluster.go b/cli/cmd/cluster.go
index 03230994..7d990284 100644
--- a/cli/cmd/cluster.go
+++ b/cli/cmd/cluster.go
@@ -26,6 +26,7 @@ import (
        "helm.sh/helm/v3/pkg/chart/loader"
        "sigs.k8s.io/yaml"

+       "github.com/kinvolk/lokomotive/pkg/assets"
        "github.com/kinvolk/lokomotive/pkg/backend"
        "github.com/kinvolk/lokomotive/pkg/backend/local"
        "github.com/kinvolk/lokomotive/pkg/components/util"
@@ -123,6 +124,14 @@ func initializeTerraform(ctxLogger *logrus.Entry, p platform.Platform, b backend
                ctxLogger.Fatalf("Failed to create Terraform executor: %v", err)
        }

+       for _, c := range p.Meta().ControlplaneCharts {
+               src := filepath.Join(assets.ControlPlaneSource, c)
+               dst := filepath.Join(assetDir, "cluster-assets", "charts", "kube-system", c)
+               if err := assets.Extract(src, dst); err != nil {
+                       ctxLogger.Fatalf("Failed to extract controlplane chart %q: %v", c, err)
+               }
+       }
+
        if err := p.Initialize(ex); err != nil {
                ctxLogger.Fatalf("Failed to initialize Platform: %v", err)
        }
diff --git a/pkg/platform/aws/aws.go b/pkg/platform/aws/aws.go
index a91ad252..ef9109cd 100644
--- a/pkg/platform/aws/aws.go
+++ b/pkg/platform/aws/aws.go
@@ -26,7 +26,6 @@ import (
        "github.com/mitchellh/go-homedir"
        "github.com/pkg/errors"

-       "github.com/kinvolk/lokomotive/pkg/assets"
        "github.com/kinvolk/lokomotive/pkg/oidc"
        "github.com/kinvolk/lokomotive/pkg/platform"
        "github.com/kinvolk/lokomotive/pkg/terraform"
@@ -119,9 +118,16 @@ func (c *config) Meta() platform.Meta {
                nodes += workerpool.Count
        }

+       charts := platform.CommonControlPlaneCharts()
+
+       if !c.DisableSelfHostedKubelet {
+               charts = append(charts, platform.KubeletChart)
+       }
+
        return platform.Meta{
-               AssetDir:      c.AssetDir,
-               ExpectedNodes: nodes,
+               AssetDir:           c.AssetDir,
+               ExpectedNodes:      nodes,
+               ControlplaneCharts: charts,
        }
 }

@@ -147,24 +153,6 @@ func (c *config) Initialize(ex *terraform.Executor) error {
                return err
        }

-       // Extract control plane chart files to cluster assets directory.
-       for _, c := range platform.CommonControlPlaneCharts {
-               src := filepath.Join(assets.ControlPlaneSource, c)
-               dst := filepath.Join(assetDir, "cluster-assets", "charts", "kube-system", c)
-               if err := assets.Extract(src, dst); err != nil {
-                       return errors.Wrapf(err, "Failed to extract charts")
-               }
-       }
-
-       // Extract self-hosted kubelet chart only when enabled in config.
-       if !c.DisableSelfHostedKubelet {
-               src := filepath.Join(assets.ControlPlaneSource, "kubelet")
-               dst := filepath.Join(assetDir, "cluster-assets", "charts", "kube-system", "kubelet")
-               if err := assets.Extract(src, dst); err != nil {
-                       return errors.Wrapf(err, "Failed to extract kubelet chart")
-               }
-       }
-
        terraformRootDir := terraform.GetTerraformRootDir(assetDir)

        return createTerraformConfigFile(c, terraformRootDir)
diff --git a/pkg/platform/baremetal/baremetal.go b/pkg/platform/baremetal/baremetal.go
index 6611b79c..9c3e3d18 100644
--- a/pkg/platform/baremetal/baremetal.go
+++ b/pkg/platform/baremetal/baremetal.go
@@ -25,7 +25,6 @@ import (
        "github.com/mitchellh/go-homedir"
        "github.com/pkg/errors"

-       "github.com/kinvolk/lokomotive/pkg/assets"
        "github.com/kinvolk/lokomotive/pkg/oidc"
        "github.com/kinvolk/lokomotive/pkg/platform"
        "github.com/kinvolk/lokomotive/pkg/terraform"
@@ -75,9 +74,16 @@ func (c *config) LoadConfig(configBody *hcl.Body, evalContext *hcl.EvalContext)

 // Meta is part of Platform interface and returns common information about the platform configuration.
 func (c *config) Meta() platform.Meta {
+       charts := platform.CommonControlPlaneCharts()
+
+       if !c.DisableSelfHostedKubelet {
+               charts = append(charts, platform.KubeletChart)
+       }
+
        return platform.Meta{
-               AssetDir:      c.AssetDir,
-               ExpectedNodes: len(c.ControllerMacs) + len(c.WorkerMacs),
+               AssetDir:           c.AssetDir,
+               ExpectedNodes:      len(c.ControllerMacs) + len(c.WorkerMacs),
+               ControlplaneCharts: charts,
        }
 }

@@ -111,24 +117,6 @@ func (c *config) Initialize(ex *terraform.Executor) error {
                return err
        }

-       // Extract control plane chart files to cluster assets directory.
-       for _, c := range platform.CommonControlPlaneCharts {
-               src := filepath.Join(assets.ControlPlaneSource, c)
-               dst := filepath.Join(assetDir, "cluster-assets", "charts", "kube-system", c)
-               if err := assets.Extract(src, dst); err != nil {
-                       return errors.Wrapf(err, "Failed to extract charts")
-               }
-       }
-
-       // Extract self-hosted kubelet chart only when enabled in config.
-       if !c.DisableSelfHostedKubelet {
-               src := filepath.Join(assets.ControlPlaneSource, "kubelet")
-               dst := filepath.Join(assetDir, "cluster-assets", "charts", "kube-system", "kubelet")
-               if err := assets.Extract(src, dst); err != nil {
-                       return errors.Wrapf(err, "Failed to extract kubelet chart")
-               }
-       }
-
        terraformRootDir := terraform.GetTerraformRootDir(assetDir)

        return createTerraformConfigFile(c, terraformRootDir)
diff --git a/pkg/platform/packet/packet.go b/pkg/platform/packet/packet.go
index 99e3d9fb..9bd1c9d6 100644
--- a/pkg/platform/packet/packet.go
+++ b/pkg/platform/packet/packet.go
@@ -29,7 +29,6 @@ import (
        "github.com/mitchellh/go-homedir"
        "github.com/pkg/errors"

-       "github.com/kinvolk/lokomotive/pkg/assets"
        "github.com/kinvolk/lokomotive/pkg/dns"
        "github.com/kinvolk/lokomotive/pkg/oidc"
        "github.com/kinvolk/lokomotive/pkg/platform"
@@ -134,9 +133,16 @@ func (c *config) Meta() platform.Meta {
                nodes += workerpool.Count
        }

+       charts := append(platform.CommonControlPlaneCharts(), "calico-host-protection")
+
+       if !c.DisableSelfHostedKubelet {
+               charts = append(charts, platform.KubeletChart)
+       }
+
        return platform.Meta{
-               AssetDir:      c.AssetDir,
-               ExpectedNodes: nodes,
+               AssetDir:           c.AssetDir,
+               ExpectedNodes:      nodes,
+               ControlplaneCharts: charts,
        }
 }

@@ -155,32 +161,6 @@ func (c *config) Initialize(ex *terraform.Executor) error {
                return err
        }

-       // Extract control plane chart files to cluster assets directory.
-       for _, c := range platform.CommonControlPlaneCharts {
-               src := filepath.Join(assets.ControlPlaneSource, c)
-               dst := filepath.Join(assetDir, "cluster-assets", "charts", "kube-system", c)
-               if err := assets.Extract(src, dst); err != nil {
-                       return errors.Wrapf(err, "Failed to extract charts")
-               }
-       }
-
-       // Extract host protection chart.
-       src := filepath.Join(assets.ControlPlaneSource, "calico-host-protection")
-       dst := filepath.Join(assetDir,
-               "cluster-assets", "charts", "kube-system", "calico-host-protection")
-       if err := assets.Extract(src, dst); err != nil {
-               return errors.Wrapf(err, "Failed to extract host protection chart")
-       }
-
-       // Extract self-hosted kubelet chart only when enabled in config.
-       if !c.DisableSelfHostedKubelet {
-               src := filepath.Join(assets.ControlPlaneSource, "kubelet")
-               dst = filepath.Join(assetDir, "cluster-assets", "charts", "kube-system", "kubelet")
-               if err := assets.Extract(src, dst); err != nil {
-                       return errors.Wrapf(err, "Failed to extract kubelet chart")
-               }
-       }
-
        terraformRootDir := terraform.GetTerraformRootDir(assetDir)

        return createTerraformConfigFile(c, terraformRootDir)
diff --git a/pkg/platform/platform.go b/pkg/platform/platform.go
index a1ba70d7..9c545f33 100644
--- a/pkg/platform/platform.go
+++ b/pkg/platform/platform.go
@@ -24,13 +24,23 @@ import (

 // CommonControlPlaneCharts defines a list of control plane Helm charts to be deployed for all
 // platforms.
-var CommonControlPlaneCharts = []string{
-       "calico",
-       "kube-apiserver",
-       "kubernetes",
-       "pod-checkpointer",
+func CommonControlPlaneCharts() []string {
+       return []string{
+               CalicoChart,
+               KubeAPIServerChart,
+               KubernetesChart,
+               PodCheckpointerChart,
+       }
 }

+const (
+       KubeletChart         = "kubelet"
+       CalicoChart          = "calico"
+       KubeAPIServerChart   = "kube-apiserver"
+       PodCheckpointerChart = "pod-checkpointer"
+       KubernetesChart      = "kubernetes"
+)
+
 // Platform describes single environment, where cluster can be installed
 type Platform interface {
        LoadConfig(*hcl.Body, *hcl.EvalContext) hcl.Diagnostics
@@ -42,9 +52,10 @@ type Platform interface {

 // Meta is a generic information format about the platform.
 type Meta struct {
-       AssetDir      string
-       ExpectedNodes int
-       Managed       bool
+       AssetDir           string
+       ExpectedNodes      int
+       Managed            bool
+       ControlplaneCharts []string
 }

 // platforms is a collection where all platforms gets automatically registered
--
2.27.0

```

Reply:
I like the idea of having `platform` only define the "contents" of a platform, however the `Meta()` method doesn't make sense to me. If we do this type of refactoring then I'd like to do it properly, namely modify the `Platform` interface to represent what a platform is - a collection of Terraform modules and Helm charts.